tests/test_sshfsmount.cpp

```diff
@@ -93,43 +94,93 @@ struct SshfsMount : public mp::test::SftpServerTest
+
+                if (found_cmd == next_expected_cmd)
                 {
+                    invoked = true;
                     output = next_expected_cmd->second;
                     remaining = output.size();
                     ++next_expected_cmd;
-                    invoked = true;
 
                     return SSH_OK;
                 }
-                else
+                else if (found_cmd != commands.end())
                 {
-                    // If not, then we have to check the rest of the remaining
-                    // commands. If the executed command is there, it means that the
-                    // executed commands do not follow the order of our expected
-                    // command vector.
-                    for (auto it = std::next(next_expected_cmd); it != commands.end(); ++it)
-                    {
-                        if (cmd == it->first)
-                        {
-                            output = it->second;
-                            remaining = output.size();
-                            ADD_FAILURE() << "\"" << (it->first) << "\" executed out of order; expected \""
-                                          << next_expected_cmd->first << "\"";
-                            return SSH_OK;
-                        }
-                    }
+                    output = found_cmd->second;
+                    remaining = output.size();
+                    ADD_FAILURE() << "\"" << (found_cmd->first) << "\" executed out of order; expected \""
+                                  << next_expected_cmd->first << "\"";
+
+                    return SSH_OK;
                 }
             }
 
+            // If the command list was entirely checked or if the executed command is not on the list, check the
+            // default commands to see if there is an answer to the current command. If not, answer with a
+            // newline, because all strings returned by the server we are mocking end in newline.
             auto it = default_cmds.find(cmd);
             if (it != default_cmds.end())
             {
                 output = it->second;
                 remaining = output.size();
                 invoked = true;
             }
+            else
+            {
+                output = "\n";
+                remaining = output.size();
+                invoked = true;
+            }
+
+            return SSH_OK;
+        };
+
+        return request_exec;
+    }
+
+    // Mock that fails after executing some commands.
+    auto make_exec_that_executes_and_fails(const CommandVector& commands, const std::string& fail_cmd,
+                                           std::string::size_type& remaining,
+                                           CommandVector::const_iterator& next_expected_cmd, std::string& output,
+                                           bool& invoked_cmd, bool& invoked_fail)
```

Comment:
Do we really need `commands` here? If `default_cmds` had all the commands necessary to go through successfully, and we'd fail on `fail_cmd`, this would get simplified a lot.

Reply:
Oh, you're right!